Continue Pi

puskin94 opened the pull request:
Hi @mintsoft and @moollaza 
I'm trying to continue the [Pi IA](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/846).
Now the situation is this:
![situa](https://cloud.githubusercontent.com/assets/8344987/6143825/6214c3b8-b1cf-11e4-8663-50a07e3baeb0.png)
The layout breaks the number... How can I fix it?

mintsoft commented:
@puskin94 Hmm, nothing jumps out immediately except that `$PI1` etc should probably explicitly be strings? (Wrap in `"` or `'`)

puskin94 commented:
@mintsoft ops, my fault. Now it should works!

puskin94 commented:
Added a simple Regex to be secure that it will only trigger numbers ;)

puskin94 commented:
@mintsoft Ok I fixed all the Bugs.... The output now is switched to a structured Answer, avoiding the broke string and beautifying the answer. Updated the test file too.
Just tell me for future improvements ;)

mintsoft commented:
@puskin94 Cool looks good, how about expanding the query space somewhat with handling for queries like `23 digits of pi` and `pi to 88 digits`?

mintsoft commented on lib/DDG/Goodie/Pi.pm:
```diff
@@ -0,0 +1,60 @@
+    my $answer;
+    my $num;
+
+    if (($decimal  > 0) && ($decimal < 400)) {
```
You could reduce the nesting here by doing this instead:
```perl
return unless $decimal  > 0 && $decimal < 400;
```

mintsoft commented on lib/DDG/Goodie/Pi.pm:
```diff
@@ -0,0 +1,60 @@
+
+    # $PI has 399 decimals
+
+    my $PI1 = '3.1415926535897932384626433832795028841971693993751058209749445923078164062862089986280348253421170679';
```
I'd probably move these outside of the handle so they don't need to be re-declared on each execution; also is there any reason why it's in 4 variables rather than one? It'd simplify the code below too if it's in one string

mintsoft commented on lib/DDG/Goodie/Pi.pm:
```diff
@@ -0,0 +1,60 @@
+# ABSTRACT: This Goodie returns 'pi' to a user-specified number of decimal places
+
+use DDG::Goodie;
+use Math::Trig;
```
I presume `Math::Trig` isn't actually need as this IA has it's own definition of `pi`?

puskin94 commented:
@mintsoft Thanks for the feedback! I hope I fixed all the stuff ;)

mintsoft commented on lib/DDG/Goodie/Pi.pm:
```diff
@@ -0,0 +1,46 @@
+description "Ex. returns 3.1415926";
+
+code_url "https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Pi.pm";
+attribution github => ["https://github.com/jmvbxx", "jmvbxx"];
```
You should add yourself here!

mintsoft commented on lib/DDG/Goodie/Pi.pm:
```diff
@@ -0,0 +1,46 @@
+
+    my $decimal = $_;
+
+    return unless $decimal && $decimal =~ /^(\d+)$/ && $decimal > 0 && $decimal < 400;
```
You could make this more dynamic by having a variable under `$PI` which is the number of characters and compare it to that? Means that if we add more digits then we don't need to update the `unless`

mintsoft commented:
@moollaza @jagtalon Think looks about ready

puskin94 commented:
@mintsoft ahah have I the privilege to change the name ?? ;)
Done !

mintsoft commented on lib/DDG/Goodie/Pi.pm:
```diff
@@ -0,0 +1,46 @@
+
+    my $decimal = $_;
+
+    return unless $decimal && $decimal =~ /^(\d+)$/ && $decimal > 0 && $decimal < length($PI);
```
I'd probably define `length($PI)`  as $PI_max_digits on L25 and then used that; that way `$PI` is only iterated over once to count the number of characters

mintsoft commented on lib/DDG/Goodie/Pi.pm:
```diff
@@ -0,0 +1,46 @@
+description "Ex. returns 3.1415926";
+
+code_url "https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Pi.pm";
+attribution github => ["https://github.com/puskin94", "puskin94"];
```
Can you add "jmvbxx" back in, you can add two github's like:
https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Conversions.pm#L18

puskin94 commented:
@mintsoft Thanks for the help with this code ;)

mintsoft commented:
@puskin94 No problem; it's what we're here for

puskin94 commented:
@mintsoft @moollaza any news guys ??

moollaza commented:
@puskin94 sorry about the delay! Will review this ASAP.

puskin94 commented:
@moollaza no problem man!! Thanks for your time ;)

moollaza commented on lib/DDG/Goodie/Pi.pm:
```diff
@@ -0,0 +1,49 @@
+    };
+
+
+    return;
```
No need for this return :) It will never be reached

moollaza commented on lib/DDG/Goodie/Pi.pm:
```diff
@@ -0,0 +1,49 @@
+
+# $PI has 399 decimals
+
+my $PI = '3.1415926535897932384626433832795028841971693993751058209749445923078164062862089986280348253421170679'.
```
Once we have Goodie templates, I think it would be cool to use a much larger number of pi digits (http://www.eveandersson.com/pi/digits/) like 1 million! And then we can easily supress the ouput from overflowing the page

puskin94 commented:
@moollaza thanks for the review!! I think that tomorrow evening you will receive the final PR ready to be merged :)

moollaza commented:
@puskin94 all that needs to be done is removing that return statement. I can do it if you'd like after merge. Its' not going to break anything :+1: 

@abeyang http://moollaza.duckduckgo.com:5000/?q=pi+150

Standard Goodie structured answer template :)

moollaza commented:
@puskin94 all that needs to be done is removing that return statement. I can do it if you'd like after merge. Its' not going to break anything :+1: 

@abeyang http://moollaza.duckduckgo.com:5000/?q=pi+150

Standard Goodie structured answer template :)

moollaza commented:
@puskin94 ok @abeyang would like a couple more updates for the UI:

- Change the `operation` to "First ### Digits of Pi", remove the `input`
- Reduce the font size of the numbers down to `1.32em`, you'll need to make a CSS file

puskin94 commented:
@moollaza hei, I can't do anything until this night! There wold be no problem for the changes to the code, but this is the first time that I add a css file to a Goodie... I hope It will be good :)

mintsoft commented:
@puskin94 Check the docs if you need any help: https://duck.co/duckduckhack/goodie_advanced_handle_functions#advanced-handle-functions

Check the "Using Files in the Share Directory"

puskin94 commented:
@mintsoft thanks man! :+1:

mintsoft commented:
@puskin94 Oops, sorry I meant: https://duck.co/duckduckhack/goodie_advanced_handle_functions#stylesheets

puskin94 commented:
@mintsoft @moollaza I think I fixed all the stuff ;) Easier than expected ;)

moollaza commented:
@puskin94 sorry I think you misunderstood me, I mean we should do the million digits later, once we have templates for Goodies, so we could allow someone to request `1000 digits of pi` but it wouldn't fill up the page like it now will. Imagine if they asked for all 1,000,000 -- you'll never see the results on the Search Page!

Also, putting a massive number like that directly into the file isn't the best way to handle it. We should have it in a separate file so that editing the Goodie itself doesn't require scroll through so many rows.

Sorry about the confusion, but for now please revert the change to 1 million digits.

puskin94 commented:
@moollaza I've done the revert! Now $PI has 399 digits ;)

moollaza commented:
@puskin94 awesome LGTM :+1: 

Design team agrees -- thanks @abeyang !

moollaza commented:
@puskin94 it's live! https://duckduckgo.com/?q=24+digits+of+pi&ia=answer

moollaza commented:
@puskin94 we now have full template support! Let's convert this one to use the `Info` template so that we can allow for larger amounts of Pi. The template will prevent the number for filling up the entire page though :)

puskin94 commented:
@moollaza sure! You should see a new PR in these days ;)
